### PR TITLE
BUG: Delegate more of Excel parsing to CSV

### LINKED ISSUE #18273: read_excel return empty dataframe when using usecols
```python
In [3]: data = pd.read_excel("A.xlsx")

In [4]: data
Out[4]:
   A  B
0  1  2
1  3  4

In [5]: data1 = pd.read_excel("A.xlsx",usecols=['B'])

In [6]: data1
Out[6]:
Empty DataFrame
Columns: []
Index: []

In [7]: pd.__version__
Out[7]: '0.21.0'

```
#### Problem description
Having a excel file name A.xlsx(or A.xls) with column A,B
read_excel return empty dataframe if usecols used

### PR DESCRIPTION
**Pull request by gfyoung**
The idea is that we read the Excel file, get the data, and let the `TextParser` handle reading and parsing, or at least most of it.  We shouldn't be doing a lot of work that is already defined in `parsers.py`

In doing so, identified several bugs:

* `index_col=None` was not being respected

* `usecols` behavior was inconsistent with that of `read_csv` for list of strings and callable inputs

* `usecols` was not being validated as proper Excel column names when passed as a string.

Closes #18273.
Closes #20480.

With regards to the latter issue, I believe this PR is a cleaner implementation that integrates the new `usecols` functionality without having to introduce a new parameter.

### CODE REVIEW

**Comment by jreback** [comment]
wow nice cleanup!

linting 

check import format using isort
ERROR: /home/travis/build/pandas-dev/pandas/pandas/io/excel.py Imports are incorrectly sorted.

File: pandas/io/excel.py [hunk 1]
```diff
@@ -642,31 +609,30 @@ def _parse_cell(cell_contents, cell_typ):
...
-                        if is_integer(skiprows):
-                            row += skiprows
-
-                        data[row], control_row = _fill_mi_header(
-                            data[row], control_row)
-                        header_name, data[row] = _pop_header_name(
-                            data[row], index_col)
-                        header_names.append(header_name)
-                else:
-                    data[header] = _trim_excel_header(data[header])
```
Comment (WillAyd) [inline pandas/io/excel.py]: Was this dead code? Maybe overlooking but don't see accounted for in the refactor

**Review by WillAyd — COMMENTED** [review]
Nice!

File: pandas/io/excel.py (hunk 1 above)
Comment (gfyoung) [reply pandas/io/excel.py]: Far from it.  Removing this is part of the bug fixing.

File: doc/source/io.rst [hunk 2]
```diff
@@ -2867,6 +2867,23 @@ indices to be parsed.
 
 Element order is ignored, so ``usecols=[0, 1]`` is the same as ``[1, 0]``.
 
+If ``usecols`` is a list of strings, it is assumed that each string corresponds
```
Comment (jreback) [inline doc/source/io.rst]: need a versionadded / changed here? or is this just added documentation?

File: pandas/io/excel.py [hunk 3]
```diff
@@ -726,6 +695,97 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    ValueError
+        Part of the Excel column name was invalid.
+    """
+    index = 0
+
+    for c in x.upper().strip():
+        cp = ord(c)
+
+        if cp < ord("A") or cp > ord("Z"):
+            raise ValueError("Invalid column name: {x}".format(x=x))
```
Comment (jreback) [inline pandas/io/excel.py]: tested?

File: pandas/io/excel.py [hunk 4]
```diff
@@ -726,6 +695,97 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    [0, 1, 2, 3, 4]
+    >>> _range2cols('A,C,Z:AB')
+    [0, 2, 25, 26, 27]
+    """
+    cols = []
+
+    for rng in areas.split(","):
+        if ":" in rng:
+            rng = rng.split(":")
+            cols += lrange(_excel2num(rng[0]), _excel2num(rng[1]) + 1)
```
Comment (jreback) [inline pandas/io/excel.py]: use .extend

File: pandas/io/excel.py [hunk 5]
```diff
@@ -726,6 +695,97 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+
+    Returns
+    -------
+    converted : object
+        The compatible format of `usecols`.
+    """
+    if usecols is None:
+        return usecols
+
+    if isinstance(usecols, int):
```
Comment (jreback) [inline pandas/io/excel.py]: use is_integer

File: pandas/tests/io/test_excel.py [hunk 6]
```diff
@@ -971,33 +1027,28 @@ def test_read_excel_chunksize(self, ext):
...
+             "date_strings": pd.date_range("2012-01-01", periods=3)})
         df2 = df.copy()
-        df2['date_strings'] = df2['date_strings'].dt.strftime('%m/%d/%Y')
+        df2["date_strings"] = df2["date_strings"].dt.strftime("%m/%d/%Y")
 
         with ensure_clean(ext) as pth:
             df2.to_excel(pth)
 
-            res = read_excel(pth)
+            res = read_excel(pth, index_col=0)
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: why is this now needed?

File: pandas/tests/io/test_excel.py [hunk 7]
```diff
@@ -1224,25 +1283,31 @@ def test_int_types(self, merge_cells, engine, ext, np_type):
     def test_float_types(self, merge_cells, engine, ext, np_type):
         # Test np.float values read come back as float.
         frame = DataFrame(np.random.random_sample(10), dtype=np_type)
-        frame.to_excel(self.path, 'test1')
+        frame.to_excel(self.path, "test1")
+
         reader = ExcelFile(self.path)
-        recons = read_excel(reader, 'test1').astype(np_type)
+        recons = read_excel(reader, "test1", index_col=0).astype(np_type)
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: why is this changed?

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. minor comments.

File: pandas/io/excel.py (hunk 3 above)
Comment (gfyoung) [reply pandas/io/excel.py]: Yes, I added a test for this (`test_usecols_excel_range_str_invalid`).

File: doc/source/io.rst (hunk 2 above)
Comment (gfyoung) [reply doc/source/io.rst]: Ah, good point.  It should have version added tags.

File: pandas/tests/io/test_excel.py (hunk 6 above)
Comment (gfyoung) [reply pandas/tests/io/test_excel.py]: `index_col=None` was not being respected before, so I need to add `index_col=0` now for this test.

File: pandas/tests/io/test_excel.py (hunk 7 above)
Comment (gfyoung) [reply pandas/tests/io/test_excel.py]: `index_col=None` was not being respected before, so I need to add `index_col=0` now for this test.

File: pandas/io/excel.py (hunk 4 above)
Comment (gfyoung) [reply pandas/io/excel.py]: Done.

File: pandas/io/excel.py (hunk 5 above)
Comment (gfyoung) [reply pandas/io/excel.py]: Done.

**Comment by gfyoung** [comment]
@jreback : Addressed comments, and all is still green.  PTAL.

File: doc/source/io.rst [hunk 8]
```diff
@@ -2867,6 +2867,27 @@ indices to be parsed.
...
+
+If ``usecols`` is a list of strings, it is assumed that each string corresponds
+to a column name provided either by the user in ``names`` or inferred from the
+document header row(s). Those strings define which columns will be parsed:
+
+.. code-block:: python
+
+    read_excel('path_to_file.xls', 'Sheet1', usecols=['foo', 'bar'])
+
+Element order is ignored, so ``usecols=['baz', 'joe']`` is the same as ``['joe', 'baz']``.
```
Comment (jreback) [inline doc/source/io.rst]: can u show an example using the excel syntax for cols as well

File: pandas/io/excel.py [hunk 9]
```diff
@@ -93,13 +92,22 @@
...
-    * If None then parse all columns,
-    * If int then indicates last column to be parsed
-    * If list of ints then indicates list of column numbers to be parsed
-    * If string then indicates comma separated list of Excel column letters and
-      column ranges (e.g. "A:E" or "A,C,E:F").  Ranges are inclusive of
+usecols : int, str, list-like, or callable default None
+    * If None, then parse all columns,
+    * If int, then indicates last column to be parsed
+    * If string, then indicates comma separated list of Excel column letters
+      and column ranges (e.g. "A:E" or "A,C,E:F"). Ranges are inclusive of
```
Comment (jreback) [inline pandas/io/excel.py]: do we have sufficient testing for ints

this last integer for a scalar seems a very odd thing to do
is this tested?

Comment (jreback) [reply pandas/io/excel.py]: should deprecate this

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/io.rst (hunk 8 above)
Comment (gfyoung) [reply doc/source/io.rst]: Done.

**Comment by gfyoung** [comment]
@jreback : Addressed comments, and all is still green. PTAL.

**Comment by gfyoung** [comment]
@jreback : Need to rebase this onto `master` because there are several places where `tm.assert_raises_regex` is used instead of `pyest.raises`.

**Comment by gfyoung** [comment]
@jreback : Comments addressed, PR rebased, and all is green.  PTAL.

File: pandas/io/excel.py (hunk 9 above)
Comment (jreback) [reply pandas/io/excel.py]: great

**Review by jreback — APPROVED** [review]
